### PR TITLE
Spatial map memory

### PR DESCRIPTION
**Pull request by aszlam**
# Description

Adds a place_field ( a 2-d grid map) object to memory, and updates it during memory.update

## Type of change

Please check the options that are relevant.

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Proposes a change (non-breaking change that isn't necessarily a bug)
- [ ] Refactor
- [x] New feature (non-breaking change that adds a new functionality)
- [ ] Breaking change (fix or feature that would break some existing functionality downstream)
- [ ] This is a unit test
- [ ] Documentation only change
- [ ] Datasets Release
- [ ] Models Release

## Type of requested review

- [x] I want a thorough review of the implementation.
- [ ] I want a high level review. 
- [ ] I want a deep design review.

# Checklist:

- [ ] I have performed manual end-to-end testing of the feature in my environment.
- [ ] I have added Docstrings and comments to the code.
- [ ] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have added relevant collaborators to review the PR before merge.
- [ ] [Polymetis only] I have verified that all scripts in `tests/scripts` runs on hardware.

### CODE REVIEW

**Comment by aszlam** [comment]
adding @soumith as a reviewer.  Can merge these in? will do next steps (with obstacle maps and visualizations on craftassist) when @anuragprat1k gets back

**Review by anuragprat1k — COMMENTED** [review]
Opened a PR (to merge into this branch) with the test scripts you can use for now at https://github.com/facebookresearch/fairo/pull/954

File: droidlet/memory/place_field.py [hunk 1]
```diff
@@ -0,0 +1,275 @@
+"""
+Copyright (c) Facebook, Inc. and its affiliates.
+"""
+
+import numpy as np
+
+MAX_MAP_SIZE = 4097
+MAP_INIT_SIZE = 1025
```
[the remark this answers is not in the document]
Comment (aszlam) [reply droidlet/memory/place_field.py:8]: just by "these seem reasonable".  can change or add to configs later

File: droidlet/memory/place_field.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+
+        self.index2memid = []
+        self.memid2index = {}
+
+        self.examined = {}
+        self.examined_id = set()
+        self.last = None
+
+        self.maps = {}
+        self.add_memid("NULL")
```
[the remark this answers is not in the document]
Comment (aszlam) [reply droidlet/memory/place_field.py]: to keep the "NULL" memid 0 in the index

File: droidlet/memory/place_field.py [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+        self.add_memid(memory.self_memid)
+        # FIXME, want slices, esp for mc... init after first perception
+        # with h=y2slice(y) instead of using 0
+        self.map_size = self.extend_map(h=0)
+
+        self.pixels_per_unit = pixels_per_unit
+
+        # gives an index allowing quick lookup by memid
+        # each entry is keyed by a memid and has nested {h: {i: {j1:True, j2:True, ...}}}
```
[the remark this answers is not in the document]
Comment (aszlam) [reply droidlet/memory/place_field.py]: I want to be able to delete and append quickly (block objects in craftassist can have thousands of covered pixels).  still you are right, it was too complicated; simplified in most recent commit

File: droidlet/memory/place_field.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+        if j == len(s) or len(levels[j]) == 0:
+            l.pop(s[j - 1])
+
+
+def no_y_l1(self, xyz, k):
+    """ returns the l1 distance between two standard coordinates"""
+    return np.linalg.norm(np.asarray([xyz[0], xyz[2]]) - np.asarray([k[0], k[2]]), ord=1)
+
+
+class PlaceField:
```
[the remark this answers is not in the document]
Comment (aszlam) [reply droidlet/memory/place_field.py:25]: added tests

**Review by anuragprat1k — APPROVED** [review]
Looks good to me. Thanks for adding tests!

**Pull request merged**